### PR TITLE
Replace callback with Promise

### PR DESCRIPTION
**Pull request by timolins**
A more modern approach, so it's possible to use Async/Await in the future. Promises are supported since Node 0.12.0, which is no longer maintained. So it's save to depend on it. The only thing here is that I use an Arrow function for context binding. They are implemented since 5.11.0, but I think we shouldn't hold on that old Node versions either.

### CODE REVIEW

**Comment by lpinca** [comment]
Thanks, I'll review this PR later. In meantime please remove the last commit. I'll handle the version when cutting the next release.

**Comment by lpinca** [comment]
Anyway Promises are not available natively in Node.js 0.12 and that version is still maintained.
That said I'm ok with dropping support for Node.js < 4.

**Comment by timolins** [comment]
Okay, I will undo the version bump when I get back to my computer. 

Seems like they are [partially supported](http://node.green/#Promise), but yeah, I still woudn't support it.

File: README.md [hunk 1]
```diff
@@ -144,27 +144,26 @@ Exchanges the authorization code for a permanent access token.
...
 
-shopifyToken.getAccessToken(hostname, code, function get(err, token) {
-  if (err) throw err;
-
+shopifyToken.getAccessToken(hostname, code).then(token => {
   console.log(token);
   // => f85632530bf277ec9ac6f649fc327f17
-});
+}).catch(err => {
+  throw err
```
Comment (lpinca) [inline README.md]: This is misleading as throwing inside a promise chain does not make the error to be thrown. Please use `console.error(err);`.

File: index.js [hunk 2]
```diff
@@ -128,79 +128,78 @@ ShopifyToken.prototype.verifyHmac = function verifyHmac(query) {
  *
  * @param {String} shop The hostname of the shop, e.g. foo.myshopify.com
  * @param {String} code The authorization code
- * @param {Function} fn Callback
- * @return {ShopifyToken} this
+ * @return {Promise} token
```
Comment (lpinca) [inline index.js]: Promise which is fulfilled with the token.

File: index.js [hunk 3]
```diff
@@ -128,79 +128,78 @@ ShopifyToken.prototype.verifyHmac = function verifyHmac(query) {
...
-  });
-
-  var timer = setTimeout(function timeout() {
-    request.abort();
-    timer = null;
-    fn(new Error('Request timed out'));
-  }, this.timeout);
+ShopifyToken.prototype.getAccessToken = function getAccessToken(shop, code) {
+  return new Promise((resolve, reject) => {
+    var data = JSON.stringify({
```
Comment (lpinca) [inline index.js]: Now that you are at it, can you please use `const` and `let`?

File: index.js [hunk 4]
```diff
@@ -128,79 +128,78 @@ ShopifyToken.prototype.verifyHmac = function verifyHmac(query) {
...
+
+        if (!timer) return;
+
+        clearTimeout(timer);
+
+        if (status !== 200) {
+          error = new Error('Failed to get Shopify access token');
+          error.responseBody = body;
+          error.statusCode = status;
+          reject(error);
```
Comment (lpinca) [inline index.js]: Missing `return` statement.

File: index.js [hunk 5]
```diff
@@ -128,79 +128,78 @@ ShopifyToken.prototype.verifyHmac = function verifyHmac(query) {
...
+        }
+
+        try {
+          body = JSON.parse(body);
+        } catch (e) {
+          error = new Error('Failed to parse the response body');
+          error.responseBody = body;
+          error.statusCode = status;
+          reject(error);
```
Comment (lpinca) [inline index.js]: Ditto.

File: test.js [hunk 6]
```diff
@@ -196,12 +196,13 @@ describe('shopify-token', function () {
         })
         .reply(200, { access_token: token });
 
-      shopifyToken.getAccessToken(hostname, code, function (err, res) {
-        if (err) return done(err);
-
+      shopifyToken.getAccessToken(hostname, code).then(function (res) {
+        console.log('123')
```
Comment (lpinca) [inline test.js]: Remove these `console.log`.
Mocha also works with promises so you can return the promise and remove the `done` callback.

File: test.js [hunk 7]
```diff
@@ -211,12 +212,12 @@ describe('shopify-token', function () {
         .post(pathname)
         .replyWithError(message);
 
-      shopifyToken.getAccessToken(hostname, '123456', function (err, res) {
+      shopifyToken.getAccessToken(hostname, '123456').catch(function (err) {
```
Comment (lpinca) [inline test.js]: You can write this as follows:
```js
return shopifyToken.getAccessToken(hostname, '123456').then(() => {
  throw new Error('Test invalidation')
}, (err) => {
  // expect ...
});
```

**Comment by lpinca** [comment]
There are some style issues, but I'll merge this when the remaining comments are addressed and fix the style nits myself.

**Comment by lpinca** [comment]
Merged as 00ebd3f805e1beb3410969618b9de4eb68e95b43.
Thanks!

**Pull request closed without merging**